[CORE-14439] `ct/l1`: stateful `compaction_committer` and forward progress on `sink/source`

WillemKauf opened the pull request:
More stuff plucked from https://github.com/redpanda-data/redpanda/compare/dev...WillemKauf:redpanda:ct_final_squashed.

This PR:
* Reworks the `sink` and `source` objects to a much better state, with clearer logic around rolling, pushing updates to the `committer`, and finalizing compaction jobs.
* Implements the `compaction_committer` as a stateful object, allowing for compaction jobs to be initialized, have newly compacted L1 objects pushed, and finalized once all the log has been compacted.

When a compaction job is initialized in the `committer`, a background fiber is kicked off which awaits newly pushed L1 objects and uploads them in the background. When the job is finalized, all inflight upload futures are awaited, and a `metastore` update is made (either through `compact_objects()`, if all uploads succeeded, or `replace_objects()`, if not all uploads succeeded). 

## Backports Required

- [X] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v25.3.x
- [ ] v25.2.x
- [ ] v25.1.x
- [ ] v24.3.x

## Release Notes

* none

WillemKauf committed:
`kafka`: abstract out `batch_from_kvs` in `produce_consume_utils`

This will be useful in more contexts as a standalone function.

WillemKauf committed:
`storage`: decompress batches in `validate` function

In case we are validating compressed batches (to avoid an assertion
failure).

WillemKauf committed:
`ct/l1/compaction`: restore `reducer_test`

A lot of code had to change, so this test was temporarily removed and
re-added with the updated interfaces.

WillemKauf committed:
`ct/l1`: add `file_and_md_info` as replacement for `object_output_t`

A more descriptive `struct` name for what will ultimately be a type
used quite frequently in the `compaction_committer`.

WillemKauf committed:
`ct/l1`: add `compaction_job_id` to `meta`

A UUID type for compaction jobs, to be used for stateful management
of compaction job states in the `compaction_committer`.

WillemKauf committed:
`ct/l1`: compile `source` and `sink` together

Unfortunately these two objects will have a dependency on each other
that is hard to break.

Compile them into one `cc_library`. This really isn't a big deal since
they will used together nearly always anyways.

WillemKauf committed:
`ct/l1`: hook `sink` back up to stateful `committer`

Hooks the `sink` up to the committer interfaces in order to:

1. Initialize a compaction job
2. Push L1 objects as they are rolled to the compaction job
3. Finalize a compaction job

WillemKauf committed:
`ct/l1`: forward progress on `source`

* Changes to internal state
* Changes to both `map_building_iteration()`:
  * `map_building_iteration` now chunks reads of dirty ranges in reads
    over the `extent`s that compromise the dirty range.
  * `map_building_reducer` and book-keeping of `new_cleaned_ranges` improved.
    The procedure for determining `new_cleaned_ranges` and whether they have tombstones
    is now done entirely in the `source`. To assist, the return type of
    `map_building_reducer::end_of_stream()` is now a `struct` containing
    a boolean indicating if the map is full (i.e if the extent was fully indexed
    before breaking out of the index loop), the `max_indexed_offset`, and
    a boolean indicating whether the range has tombstones or not.
* Changes to `deduplication_iteration()`:
  * `min.compaction.lag.ms` is now supported, due to use of helper function
    `should_compact_extent()`.

WillemKauf committed:
`ct/l1`: add new state to `sink` and fill out `initialize()` details

Reworking of the internal state of `sink` and setting up the `initialize()`
function call with the necessary shared state from the `source`.

WillemKauf committed:
`ct/l1`: implement stateful `compaction_committer`

Provides interfaces for:

1. Initializing a compaction job, which spins off a background
   fiber which will take new updates and push them to cloud storage
   while building up a `object_metadata_builder` update for the `metastore`.
2. Pushing updates to the `compaction_committer` for a given `compaction_job_id`.
3. Finalizing a `compaction_job_id`, which will force the background fiber
   to finish its uploads, build a `metastore::compaction_map_t` update, and
   finalize a commit to the `metastore`.

A background fiber is kicked off for every `compaction_job_id`, but it is
currently expected that only one compaction job should be in progress per
shard at a time.

WillemKauf committed:
`ct/l1`: forward progress on `sink`

Implements rolling, setting extent bounds, and determining removed tombstone
ranges and dirty ranges made clean during finalization.

WillemKauf commented on src/v/cloud_topics/level_one/compaction/tests/reducer_test.cc:
```diff
@@ -84,26 +168,85 @@ TEST(ReducerTest, InMemoryReducer) {
+                auto key = iobuf_to_string(rec.release_key());
+                std::optional<ss::sstring> val;
+                if (rec.has_value()) {
+                    auto val = iobuf_to_string(rec.release_value());
```
Nice catch mr robot

WillemKauf commented on src/v/cloud_topics/level_one/compaction/committer.cc:
```diff
@@ -22,89 +29,403 @@ compaction_committer::compaction_committer(
+
+ss::future<> compaction_committer::cancel_active_jobs() {
+    static constexpr size_t max_concurrent_removal = 1024;
+    for (auto& [job, state] : _compaction_jobs) {
```
also valid

andrwng commented on src/v/cloud_topics/level_one/compaction/source.cc:
```diff
@@ -32,14 +32,17 @@ namespace {
-        bool map_is_full{false};
+    struct return_t {
+        bool map_is_full;
+        std::optional<model::offset> max_indexed_offset;
```
nit: since we're still in the cloud topics namespace, can we return optional<kafka::offset> instead and do whatever casting we need closer to where the kafka<->model switch happens? Or does that complicate other things?

WillemKauf replied:
It feels simpler to leave things in `map_building_reducer` as `model::offset`, and then cast back when dealing with its return values. Some of the code in `maybe_index_record_in_map()` is a bit uglier when doing the casting locally.

WillemKauf commented:
Force push to make changes suggested by our robot overlords

andrwng commented on src/v/cloud_topics/level_one/compaction/source.cc:
```diff
@@ -122,103 +176,114 @@ ss::future<ss::stop_iteration> compaction_source::map_building_iteration() {
+    auto dirty_range_intervals = intervals_for_dirty_range(
+      dirty_range, _extents);
+    bool map_is_full = false;
+    for (auto it = dirty_range_intervals.crbegin();
```
nit: can you add a comment explaining why we're iterating backwards, either here or by the iterator definition?

andrwng commented on src/v/cloud_topics/level_one/compaction/source.cc:
```diff
@@ -122,103 +176,114 @@ ss::future<ss::stop_iteration> compaction_source::map_building_iteration() {
+    const auto& dirty_range = *_dirty_range_it;
+
+    auto dirty_range_intervals = intervals_for_dirty_range(
+      dirty_range, _extents);
```
Let's call this something else. It is very easy to confuse this with _dirty_range_intervals. How about extent_aligned_ranges or something

WillemKauf replied:
Done

andrwng commented on src/v/cloud_topics/level_one/compaction/source.cc:
```diff
@@ -32,14 +32,17 @@ namespace {
-        bool map_is_full{false};
+    struct return_t {
+        bool map_is_full;
+        std::optional<model::offset> max_indexed_offset;
```
I'm a little nervous of this returning a single offset, vs returning the offset interval set representing the offsets added to the map

WillemKauf commented on src/v/cloud_topics/level_one/compaction/tests/reducer_test.cc:
```diff
@@ -84,139 +84,3 @@ TEST(ReducerTest, InMemoryReducer) {
       std::move(output_batches));
 }
-
-class ReducerTestFixture : public l1::l1_reader_fixture {
```
This test is coming back in a future commit, don't worry!

andrwng commented on src/v/cloud_topics/level_one/compaction/source.h:
```diff
@@ -40,38 +46,45 @@ class compaction_source : public compaction::sliding_window_reducer::source {
+    const offset_interval_set& _removable_tombstone_ranges;
+
+    // Iterator used during `map_building_iteration()` which points into the
+    // above vector `_dirty_range_intervals`.
```
nit: probably worth a comment explaining why we're iterating backwards

WillemKauf replied:
Added

andrwng commented on src/v/cloud_topics/level_one/compaction/source.h:
```diff
@@ -40,38 +46,45 @@ class compaction_source : public compaction::sliding_window_reducer::source {
+    // The extents of the CTP, as returned from the `metastore`.
+    metastore::extent_metadata_vec _extents;
+    metastore::extent_metadata_vec::const_iterator _extents_it;
+    metastore::extent_metadata_vec::const_iterator _extents_end_it;
```
Do you expect that this will always be all of the partition's extents? If so, do you have longer term ideas for making that not the case? This seems like something we might want to change sooner rather than later, and I'm wondering if it's a big lift (and either way, whether it's easier to make that lift now or later)

WillemKauf replied:
>Do you expect that this will always be all of the partition's extents

Currently, that is the assumption in the code, yes.

>If so, do you have longer term ideas for making that not the case?

If you are referring to the potential issue of not being able to return all the extents in the metastore in one RPC response and instead needing to paginate `extent_metadata`, I think this will be an easy enough fix with the right abstraction here (something similar to an iterator or a stream, but may from time to time block on an RPC call to get more `extent_metadata`s from the `metastore`).

I think this is probably a good thing to tackle sooner rather than later.

andrwng commented on src/v/cloud_topics/level_one/compaction/source.cc:
```diff
@@ -122,103 +176,114 @@ ss::future<ss::stop_iteration> compaction_source::map_building_iteration() {
-          _ntp,
-          stats);
+    auto& extent = *_extents_it;
+    if (should_compact_extent(extent, _min_compaction_lag_ms)) {
```
Should we stop iterating if we get to an extent that shouldn't be compacted? Since each pass may remove tombstones, and there's no relationship between min lag and tombstone removal delay, is it possible that we'll end up incorrectly removing a tombstone in a later extent?

WillemKauf replied:
> Should we stop iterating if we get to an extent that shouldn't be compacted?

I don't think so. Kafka (in the traditional sense) does not perscribe that segments in the log need be monotonically increasing w/r/t max timestamp, and the same should be true of extents, meaning some earlier extents/segments may be ineligible for compaction, while some newer ones are.

> is it possible that we'll end up incorrectly removing a tombstone in a later extent?

I don't quite see how this is possible, or how a tombstone could be "incorrectly" removed. Tombstones don't need to be removed in any sort of order in the log, if they are eligible for deletion they can be removed unconditionally. As you know, the `metastore` response dictates which regions of the log can have their tombstones removed, and any tombstone in those regions can be "correctly" removed.

andrwng replied:
Ah fair thanks, I momentarily forgot how the metastore compaction metadata worked :)

andrwng commented on src/v/cloud_topics/level_one/compaction/source.cc:
```diff
@@ -122,103 +176,114 @@ ss::future<ss::stop_iteration> compaction_source::map_building_iteration() {
+              "stats: {}",
+              start_offset,
+              last_offset,
+              _ntp,
```
nit: not from this PR, but the arguments to this vlog look out of order

WillemKauf replied:
Ah thanks, bad rebase

andrwng commented on src/v/cloud_topics/level_one/compaction/source.cc:
```diff
@@ -122,103 +176,114 @@ ss::future<ss::stop_iteration> compaction_source::map_building_iteration() {
+    bool map_is_full = false;
+    for (auto it = dirty_range_intervals.crbegin();
+         it != dirty_range_intervals.crend();
+         ++it) {
```
nit: I think it's worth commenting somewhere that this alignment to extents isn't required for correctness, but serves the purposes of:
- giving us natural chunks to iterate backwards on, and
- giving us naturally fine-grained intervals to indicate the presence of tombstones

I think an alternate approach would have been to iterate backwards with some fixed chunk size and it would have been equally correct. That might not be a bad idea either, if we end up having to not rely on depending on so much extent information for scalability reasons


On the flip side, I think it's also worth commenting in the forwards pass that the extent alignment is required as a part of the contract with the metastore

WillemKauf replied:
Added

andrwng commented on src/v/cloud_topics/level_one/compaction/source.cc:
```diff
@@ -249,6 +249,8 @@ ss::future<ss::stop_iteration> compaction_source::deduplication_iteration(
+        co_await ct_sink.process_next_extent_offset_bounds(
+          start_offset, last_offset);
         auto stats = co_await rdr.consume(
           compaction_filter{ct_sink, *_map, _ntp, _removable_tombstone_ranges},
```
Idea: make the ct_sink get this from the reader somehow, since the reader has the start and last offset?

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--2132b028-0313-460c-9b28-d1f06118e8ec-->
<!--TEST RESULTS-->
<!--e10bedc6-0e20-4e5a-82ec-973e471e2619-->
<!--TEST RESULTS-->
<!--02ea5111-39eb-4c81-81b5-d727ab56d0bf-->
<!--TEST RESULTS-->
<!--6869be9e-9a5c-4525-9416-40541dba6254-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/77300">build#77300</a></summary>


| test_class               | test_method                           | test_arguments                    | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                | test_history                                                                                                                                                                 |
|:-------------------------|:--------------------------------------|:----------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:----------------------------------------------------------------------------------------------------------------------------------------------------------------------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| NodesDecommissioningTest | test_decommissioning_rebalancing_node | {"shutdown_decommissioned": true} | integration | https://buildkite.com/redpanda/redpanda/builds/77300#019ae6fc-2f58-47e3-9baa-c6be088c5ca5 | FLAKY         | 18/21    | upstream reliability is '94.1304347826087'. current run reliability is '85.71428571428571'. drift is 8.41615 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=NodesDecommissioningTest&test_method=test_decommissioning_rebalancing_node |
| SimpleEndToEndTest       | test_relaxed_acks                     | {"write_caching": false}          | integration | https://buildkite.com/redpanda/redpanda/builds/77300#019ae6fc-2f5e-4e55-8639-8013f1434052 | FLAKY         | 19/21    | upstream reliability is '98.78048780487805'. current run reliability is '90.47619047619048'. drift is 8.3043 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=SimpleEndToEndTest&test_method=test_relaxed_acks                           |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/77313">build#77313</a></summary>


| test_class                          | test_method    | test_arguments              | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 | test_history                                                                                                                                                     |
|:------------------------------------|:---------------|:----------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------|
| WriteCachingFailureInjectionE2ETest | test_crash_all | {"use_transactions": false} | integration | https://buildkite.com/redpanda/redpanda/builds/77313#019ae7c7-801d-42db-bf60-815581943635 | FLAKY         | 16/21    | upstream reliability is '89.2638036809816'. current run reliability is '76.19047619047619'. drift is 13.07333 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=WriteCachingFailureInjectionE2ETest&test_method=test_crash_all |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/77744">build#77744</a></summary>


| test_class                       | test_method                | test_arguments                                                            | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                                                                                                                                                                                                                                              | test_history                                                                                                                                                              |
|:---------------------------------|:---------------------------|:--------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:--------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ReplicatedMetastoreTest          | TestBasicRemoveTopics      |                                                                           | unit        | https://buildkite.com/redpanda/redpanda/builds/77744#019b0f58-52aa-4cf4-8fe1-e659c6ac174b | FAIL          | 0/1      |                                                                                                                                                                                                                                                                                                                                                                                                     |                                                                                                                                                                           |
| FeaturesMultiNodeUpgradeTest     | test_rollback              | null                                                                      | integration | https://buildkite.com/redpanda/redpanda/builds/77744#019b0f87-225b-4930-bac2-9a71aed5ce9a | FAIL          | 0/11     | Test FAILS after retries.[Significant increase in flaky rate](https://redpandadata.atlassian.net/wiki/spaces/DEVPROD/pages/1419247661/Statistical+retry+failure+reasons#%5BinlineExtension%5DCase-B%3A-Test-fails-with-an-improbably-higher-rate-than-baseline-(run-fails))(baseline=0.0882, p0=0.0000, reject_threshold=0.0100)                                                                    | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=FeaturesMultiNodeUpgradeTest&test_method=test_rollback                  |
| FeaturesMultiNodeUpgradeTest     | test_upgrade               | null                                                                      | integration | https://buildkite.com/redpanda/redpanda/builds/77744#019b0f87-225c-4685-b12e-b7e2d978e557 | FAIL          | 0/11     | Test FAILS after retries.[Significant increase in flaky rate](https://redpandadata.atlassian.net/wiki/spaces/DEVPROD/pages/1419247661/Statistical+retry+failure+reasons#%5BinlineExtension%5DCase-B%3A-Test-fails-with-an-improbably-higher-rate-than-baseline-(run-fails))(baseline=0.0882, p0=0.0000, reject_threshold=0.0100)                                                                    | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=FeaturesMultiNodeUpgradeTest&test_method=test_upgrade                   |
| FeaturesNodeJoinTest             | test_old_node_join         | null                                                                      | integration | https://buildkite.com/redpanda/redpanda/builds/77744#019b0f87-225e-4575-9177-8a147616dd38 | FAIL          | 0/11     | Test FAILS after retries.[Significant increase in flaky rate](https://redpandadata.atlassian.net/wiki/spaces/DEVPROD/pages/1419247661/Statistical+retry+failure+reasons#%5BinlineExtension%5DCase-B%3A-Test-fails-with-an-improbably-higher-rate-than-baseline-(run-fails))(baseline=0.1146, p0=0.0000, reject_threshold=0.0100)                                                                    | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=FeaturesNodeJoinTest&test_method=test_old_node_join                     |
| FeaturesSingleNodeUpgradeTest    | test_upgrade               | null                                                                      | integration | https://buildkite.com/redpanda/redpanda/builds/77744#019b0f87-2255-4210-a445-24fc1df52bf3 | FAIL          | 0/11     | Test FAILS after retries.[Significant increase in flaky rate](https://redpandadata.atlassian.net/wiki/spaces/DEVPROD/pages/1419247661/Statistical+retry+failure+reasons#%5BinlineExtension%5DCase-B%3A-Test-fails-with-an-improbably-higher-rate-than-baseline-(run-fails))(baseline=0.1146, p0=0.0000, reject_threshold=0.0100)                                                                    | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=FeaturesSingleNodeUpgradeTest&test_method=test_upgrade                  |
| LicenseEnforcementTest           | test_license_enforcement   | {"clean_node_after_recovery": false, "clean_node_before_recovery": false} | integration | https://buildkite.com/redpanda/redpanda/builds/77744#019b0f87-225a-446e-bef2-119b5b3de0e4 | FAIL          | 0/11     | Test FAILS after retries.[Significant increase in flaky rate](https://redpandadata.atlassian.net/wiki/spaces/DEVPROD/pages/1419247661/Statistical+retry+failure+reasons#%5BinlineExtension%5DCase-B%3A-Test-fails-with-an-improbably-higher-rate-than-baseline-(run-fails))(baseline=0.1014, p0=0.0000, reject_threshold=0.0100)                                                                    | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=LicenseEnforcementTest&test_method=test_license_enforcement             |
| LicenseEnforcementTest           | test_license_enforcement   | {"clean_node_after_recovery": true, "clean_node_before_recovery": false}  | integration | https://buildkite.com/redpanda/redpanda/builds/77744#019b0f87-225b-4930-bac2-9a71aed5ce9a | FAIL          | 0/11     | Test FAILS after retries.[Significant increase in flaky rate](https://redpandadata.atlassian.net/wiki/spaces/DEVPROD/pages/1419247661/Statistical+retry+failure+reasons#%5BinlineExtension%5DCase-B%3A-Test-fails-with-an-improbably-higher-rate-than-baseline-(run-fails))(baseline=0.1014, p0=0.0000, reject_threshold=0.0100)                                                                    | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=LicenseEnforcementTest&test_method=test_license_enforcement             |
| LicenseEnforcementTest           | test_license_enforcement   | {"clean_node_after_recovery": false, "clean_node_before_recovery": true}  | integration | https://buildkite.com/redpanda/redpanda/builds/77744#019b0f87-225c-4685-b12e-b7e2d978e557 | FAIL          | 0/11     | Test FAILS after retries.[Significant increase in flaky rate](https://redpandadata.atlassian.net/wiki/spaces/DEVPROD/pages/1419247661/Statistical+retry+failure+reasons#%5BinlineExtension%5DCase-B%3A-Test-fails-with-an-improbably-higher-rate-than-baseline-(run-fails))(baseline=0.1014, p0=0.0000, reject_threshold=0.0100)                                                                    | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=LicenseEnforcementTest&test_method=test_license_enforcement             |
| LicenseEnforcementTest           | test_license_enforcement   | {"clean_node_after_recovery": true, "clean_node_before_recovery": true}   | integration | https://buildkite.com/redpanda/redpanda/builds/77744#019b0f87-225e-4575-9177-8a147616dd38 | FAIL          | 0/11     | Test FAILS after retries.[Significant increase in flaky rate](https://redpandadata.atlassian.net/wiki/spaces/DEVPROD/pages/1419247661/Statistical+retry+failure+reasons#%5BinlineExtension%5DCase-B%3A-Test-fails-with-an-improbably-higher-rate-than-baseline-(run-fails))(baseline=0.1014, p0=0.0000, reject_threshold=0.0100)                                                                    | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=LicenseEnforcementTest&test_method=test_license_enforcement             |
| WriteCachingFailureInjectionTest | test_unavoidable_data_loss | null                                                                      | integration | https://buildkite.com/redpanda/redpanda/builds/77744#019b0f87-225c-4685-b12e-b7e2d978e557 | FLAKY         | 18/21    | Test PASSES after retries.[No significant increase in flaky rate](https://redpandadata.atlassian.net/wiki/spaces/DEVPROD/pages/1419247661/Statistical+retry+failure+reasons#%5BinlineExtension%5DCase-C%3A-Test-passes-with-a-rate-not-significantly-worse-than-baseline-(run-passes))(baseline=0.0561, p0=0.3102, reject_threshold=0.0100. adj_baseline=0.1590, p1=0.3625, trust_threshold=0.5000) | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=WriteCachingFailureInjectionTest&test_method=test_unavoidable_data_loss |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/77772">build#77772</a></summary>


| test_class                 | test_method               | test_arguments                                                                                                                        | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                                                                                                                                                                                                                                              | test_history                                                                                                                                                       |
|:---------------------------|:--------------------------|:--------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:-------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ReplicatedMetastoreTest    | TestBasicRemoveTopics     |                                                                                                                                       | unit        | https://buildkite.com/redpanda/redpanda/builds/77772#019b1319-bcf5-4e94-8580-5a3443dfc982 | FAIL          | 0/1      |                                                                                                                                                                                                                                                                                                                                                                                                     |                                                                                                                                                                    |
| PartitionReassignmentsTest | test_reassignments_cancel | null                                                                                                                                  | integration | https://buildkite.com/redpanda/redpanda/builds/77772#019b138b-5d65-4264-85c0-590f3ab9ed90 | FLAKY         | 9/11     | Test PASSES after retries.[No significant increase in flaky rate](https://redpandadata.atlassian.net/wiki/spaces/DEVPROD/pages/1419247661/Statistical+retry+failure+reasons#%5BinlineExtension%5DCase-C%3A-Test-passes-with-a-rate-not-significantly-worse-than-baseline-(run-passes))(baseline=0.0815, p0=0.5726, reject_threshold=0.0100. adj_baseline=0.2251, p1=0.3049, trust_threshold=0.5000) | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=PartitionReassignmentsTest&test_method=test_reassignments_cancel |
| TieredStorageTest          | test_tiered_storage       | {"cloud_storage_type_and_url_style": [2, "virtual_host"], "test_case": {"name": "(TS_Read == True, SegmentRolledByTimeout == True)"}} | integration | https://buildkite.com/redpanda/redpanda/builds/77772#019b138b-5d61-4b7f-9dd8-bff7caa3a9cc | FLAKY         | 50/51    | Test PASSES after retries.[Inconclusive result after max retries](https://redpandadata.atlassian.net/wiki/spaces/DEVPROD/pages/1419247661/Statistical+retry+failure+reasons#Case-F%3A-inconclusive-result-after-max-retries-(run-passes)%5BinlineExtension%5D)(baseline=0.0009, p0=1.0000, reject_threshold=0.0100. adj_baseline=0.0028, p1=0.8672, trust_threshold=0.5000)                         | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=TieredStorageTest&test_method=test_tiered_storage                |
</details>
<!--/body-->

andrwng commented on src/v/cloud_topics/level_one/compaction/sink.cc:
```diff
@@ -162,20 +145,118 @@ ss::future<> compaction_sink::maybe_roll() {
+    auto prev_offset = std::max(
+      kafka::prev_offset(model::offset_cast(b.base_offset())),
+      kafka::offset{0});
+    set_last_processed_offset(prev_offset);
```
Why are we setting the prev up front, rather setting the last on exit? A comment would be nice if setting the last offset is not correct

WillemKauf replied:
We have to set the `_last_processed_offset` _before_ calling into `maybe_roll()` in order to get the offset bounds of the built L1 object correct.

I.e, imagine a bunch of batches between `[0] [2], [3], [5], [10]` have been removed, and the currently active `L1` object needs a roll when batch`[10]` is passed to `sink::operator()`. Technically that `L1` object spans batches `[0] -> [9]`, but we are otherwise unaware of this because they have not been supplied to the `sink`.

We could also call `set_last_processed_offset()` _after_ adding the batch, but it doesn't help us for the reason stated above (you'll notice `set_last_processed_offset()` is usually called before every call to `roll()` or `maybe_roll()`.)

andrwng commented on src/v/cloud_topics/level_one/compaction/sink.h:
```diff
@@ -39,15 +39,38 @@ class compaction_sink : public compaction::sliding_window_reducer::sink {
+    // is non-contiguous. For example, if the extents were
+    // `[[0,10],[11,20],[21,30]]`, and the extent [11,20] was deemed ineligible
+    // for compaction (due to `min.compaction.lag.ms` or some other reason), the
+    // current L1 object composing the range [0,10] would be rolled, and a new
+    // L1 object would be started for the range [21,30].
```
See my other question about whether we should stop after seeing an ineligible extent. It certainly seems simpler (and conceptually easier to reason about) if we ensure the forwards pass is done in order

WillemKauf replied:
Please see my other response on Kafka semantics and compaction. I don't think it make senses to potentially block all of compaction due to one segment or extent with an ineligible timestamp.

andrwng commented on src/v/cloud_topics/level_one/compaction/sink.h:
```diff
@@ -46,28 +49,59 @@ class compaction_sink : public compaction::sliding_window_reducer::sink {
     // Guaranteed to have a value iff _active_staging_file.
     std::unique_ptr<object_builder> _builder{nullptr};
+
+    // The current `_extent`'s offsets. Batches received by the sink's
```
Is `_extent` a variable? I'm missing it

WillemKauf replied:
Oops, nope, bit rot

andrwng reviewed (commented):
Haven't gotten to the compaction committer yet, but will look more later/tomorrow

WillemKauf commented:
Force push to:
* Add requested comments
* Fix order of arguments in `vlog` statement
* rename `dirty_range_intervals` -> `extent_aligned_ranges`

andrwng commented on src/v/cloud_topics/level_one/compaction/sink.cc:
```diff
@@ -162,20 +145,119 @@ ss::future<> compaction_sink::maybe_roll() {
+
+ss::future<> compaction_sink::roll(bool initialize_new_builder) {
+    // 1. Push currently built L1 object & metadata to the committer.
+    if (_active_staging_file) {
```
I think we are fundamentally keeping track of two things:
- the currently pending object
- the total set of input extents that we have processed

We track these to achieve a few goals:
- trim our final resulting metadata (cleaned ranges, removed tombstone ranges) to be a subset of the processed input extents
- make sure the objects that we're building are contiguous when the inputs are contiguous
- make sure we start new objects when inputs are not contiguous

If these seem correct, I think it's worth documenting these in a comment somewhere, maybe above the members or in the header above the class declaration. They're pretty nuanced, and without this knowledge it's hard to understand why the code is doing what it is.

andrwng replied:
With those goals in mind, I'm thinking if we can simplify the amount of state tracked in the sink. WDYT about adding a new struct to encapsulate the in-progress output object?

```
struct compacted_object {
  staging_file active_staging_file;
  object_builder builder;
  kafka::offset base_offset;
};

std::unique_ptr<compacted_object> inflight_object;
std::optional<kafka::offset> iteration_base_offset; // set only if we're in the middle of iterating
offset_interval_set iterated_extents; // instead of processed extents?
```

I think `_last_processed_offset` is mostly there to be able to preserve that the output objects are contiguous when we're iterating. I think that can instead be accomplished by adding an argument to roll, like `roll(kafka::offset last_offset)`, where in the `operator()` we pass it `prev(batch.base_offset)`.

I think the dance around keeping the _next_ extents confusing, especially given the nuance with the first object. I'm thinking it may make sense to split up `process_next_extent_offset_bounds()` into `prepare_iteration(kafka::offset base_offset)` and `finish_iteration(kafka::offset last_offset)`, where:
- prepare sets `iteration_base_offset` and initializes the `inflight_object` if it doesn't exist, and if called with a base offset that doesn't align with the last processed offset, it calls `roll(_processed_calls.back().last_offset)`
- the finish calls `_processed_calls.insert(extent_base_offset, last_offset)`, and is called after the consume call finishes

TLDR: I think it isn't too unnatural to have iteration bounds be a part of the sink's interface, and having an in-flight object also feels natural. The details, I could be way off base, so maybe it breaks down somewhere. WDYT?

WillemKauf replied:
Done. We now have `prepare_iteration` and `finish_iteration` calls around the processing of batches for the `sink`.
This also helps with reducing the amount of book-kept state within `sink` - `_extent_base_offset` and `_extent_last_offset` are gone, and `roll()` now takes two arguments: `roll(object_last_offset, next_base_offset)`. These are two separate offsets because there is no guarantee that they are contiguous.

andrwng commented on src/v/cloud_topics/level_one/compaction/sink.cc:
```diff
@@ -162,20 +145,119 @@ ss::future<> compaction_sink::maybe_roll() {
+    if (is_first_extent) {
+        _object_base_offset = next_extent_base;
+    } else {
+        _processed_extents.insert(_extent_base_offset, _extent_last_offset);
```
These are input extent offsets

andrwng replied:
I think we are fundamentally keeping track of two things:
- the currently pending object
- the total set of input extents that we have seen

andrwng commented on src/v/cloud_topics/level_one/compaction/sink.cc:
```diff
@@ -30,6 +30,43 @@
 
 namespace cloud_topics::l1 {
 
+namespace {
+
+offset_interval_set get_removed_tombstone_ranges(
+  const offset_interval_set& removable_tombstone_ranges,
+  const offset_interval_set& processed_extents) {
+    offset_interval_set removed_tombstone_ranges;
+    auto stream = removable_tombstone_ranges.make_stream();
+    while (stream.has_next()) {
+        auto i = stream.next();
+        if (processed_extents.covers(i.base_offset, i.last_offset)) {
+            removed_tombstone_ranges.insert(i.base_offset, i.last_offset);
+        }
+    }
+    return removed_tombstone_ranges;
+}
+
+chunked_vector<metastore::compaction_update::cleaned_range>
+get_new_cleaned_ranges(
+  const chunked_vector<metastore::compaction_update::cleaned_range>&
+    maybe_cleaned_ranges,
+  const offset_interval_set& processed_extents) {
+    chunked_vector<metastore::compaction_update::cleaned_range>
+      new_cleaned_ranges;
+    new_cleaned_ranges.reserve(maybe_cleaned_ranges.size());
+    for (const auto& cleaned_range : maybe_cleaned_ranges) {
+        if (processed_extents.covers(
+              cleaned_range.base_offset, cleaned_range.last_offset)) {
+            new_cleaned_ranges.push_back(cleaned_range);
+        }
+    }
+
+    new_cleaned_ranges.shrink_to_fit();
+    return new_cleaned_ranges;
+}
```
nit: please add comments explaining why these are needed

WillemKauf replied:
Added!

andrwng commented on src/v/cloud_topics/level_one/compaction/committer.cc:
```diff
@@ -22,89 +29,403 @@ compaction_committer::compaction_committer(
+      ssx::when_all_succeed<res_t>(std::move(inflight_uploads)));
+    vassert(
+      !res.failed(),
+      "expected that these futures wouldn't throw, just propagate an error.");
```
Can't they throw if the job mutex is broken? What prevents that?

andrwng commented on src/v/cloud_topics/level_one/compaction/committer.cc:
```diff
@@ -22,89 +29,403 @@ compaction_committer::compaction_committer(
+    finalize_op op = finalize_op::compact_objects;
+    if (fut.failed()) {
+        auto e = fut.get_exception();
+        auto lvl = ssx::is_shutdown_exception(e) ? ss::log_level::debug
```
At what point do we stop if we're shutting down? I'm kind of surprised to see us continue on for shutdown errors

andrwng commented on src/v/cloud_topics/level_one/compaction/committer.cc:
```diff
@@ -22,89 +29,403 @@ compaction_committer::compaction_committer(
+}
+
+ss::future<compaction_job_id>
+compaction_committer::begin_compaction_job(model::topic_id_partition tidp) {
```
Maybe vassert that _compaction_jobs is empty? (and also comment where to look to verify how that is a strict guarantee)

WillemKauf committed:
`ct/l1`: fix start offset bug in `simple_metastore::get_dirty_ratio()`

Don't consider any parts of partially truncated extents in the calculation.

WillemKauf commented:
Converting to draft mode, not because I don't think this PR isn't ready for review, but because I think it will go through a few more iterations and there's no point in wasting money on CI runs.

andrwng commented on src/v/cloud_topics/level_one/compaction/sink.cc:
```diff
@@ -73,109 +121,171 @@ compaction_sink::initialize(compaction::sliding_window_reducer::source& src) {
+    if (needs_roll()) {
+        co_await roll(last_offset, next_offset);
+    }
 }
```
nit: roll, maybe_roll, and needs_roll all seem to be used just once or twice. Now that flush() is its own thing, I think it'll be easier to follow and maintain if we just inline calls to flush() and initialize_builder().

WillemKauf replied:
Done. Removed `roll()` and `maybe_roll()`, left `needs_roll()`, inlined calls to `flush()` and `initialize_builder()`.

andrwng replied:
What's the rationale for keeping needs_roll()? The other rolling terminology is gone, it's only used once, and it also seems easy to inline?

andrwng replied:
If the concern is it being awkward to call flush() when conditioning on _inflight_object, maybe write it as:

```
if (_inflight_object && _inflight_object->size > target) {
  co_await flush(prev_offset);
}
if (!_inflight_object) {
  co_await initialize_builder(next_offset);
}
```

WillemKauf replied:
Done

andrwng commented on src/v/cloud_topics/level_one/compaction/sink.cc:
```diff
@@ -73,109 +121,171 @@ compaction_sink::initialize(compaction::sliding_window_reducer::source& src) {
+        co_await initialize_builder(next_extent_base);
+    } else {
+        auto prev_extent_last_offset
+          = _processed_extents.to_vec().back().last_offset;
```
to_vec() materializes the vector, which is a bit much if our goal is to just get the last offset. It's small, but might be worth considering exposing last_offset of offset_interval_set, or having some reverse-stream function, or exposing the back interval instead.

WillemKauf replied:
Exposed reverse iterators in `interval_set` and `offset_interval_set` via a template on `offset_interval_set::stream`, and added `offset_interval_set::make_reverse_stream()`.

andrwng commented on src/v/cloud_topics/level_one/compaction/sink.cc:
```diff
@@ -73,109 +121,171 @@ compaction_sink::initialize(compaction::sliding_window_reducer::source& src) {
+ss::future<>
+compaction_sink::prepare_iteration(kafka::offset next_extent_base) {
+    bool is_first_extent = _processed_extents.empty();
+    if (is_first_extent) {
```
nit: if we go the route of splitting up roll into flush() and initalize_builder(), I think this can be tweaked a bit to feel less special casey:

```
if (!_processed_extents.empty()) {
  if (next_extent_base == next(prev_extent_last_offset)) {
    co_return;
  }
  // Passed extents are non-contiguous...
  co_await flush(prev_extent_last_offset);
  // Intentional fallthrough.
}
co_await initialize_builder(next_extent_base);
```

I think then it'll be a bit more obvious that if we're initializing a builder, it's always aligned with next_extent_base

WillemKauf replied:
Done.

andrwng commented on src/v/cloud_topics/level_one/compaction/committer.h:
```diff
@@ -27,61 +32,204 @@ class compaction_committer {
+      begin_compaction_job(model::topic_id_partition);
+
+    // Adds a newly produced L1 object and its related object info to the
+    // in-progress state for the given compaction job.
```
nit: if we're exposing "jobs" as a primitive anyway via compaction job ID, maybe it makes sense to have the committer to pass a pointer to the job instead, and have callers call some `job.add_l1_object()` and `finalize()`?

I'm leaning this way because it seems like a lot of the logic currently in the committer is actually already encapsulated by the job, and I think it'd make for cleaner/more testable code if committer-specific logic (orchestrating several jobs) and job-specific logic (the encapsulated bits for IO and building updates) were separate, especially if we expect to make the committer manage several jobs later. WDYT?

WillemKauf replied:
I'll think about this more and come back to it.

WillemKauf replied:
I heavily reworked the `compaction_committer` implementation details in this commit: https://github.com/redpanda-data/redpanda/pull/28831/commits/206b268e5332986221ab8a56729fafab13549cce.

It may be useful to look at the changed `commiter.cc/.h` files in isolation instead of the diff. If you like the direction of the work, I'll squash the changes into the original commit eventually.

WillemKauf committed:
`container`: add reverse iterators to `interval_set`

WillemKauf committed:
`ct/l1`: add `offset_interval_set::make_reverse_stream()`

WillemKauf commented on src/v/cloud_topics/level_one/compaction/committer.cc:
```diff
@@ -22,89 +29,432 @@ compaction_committer::compaction_committer(
+      compaction_update_str);
+}
+
+ss::future<> compaction_committer::do_replace_objects(job_ptr_t job) {
```
Per our discussions @andrwng, this should probably become a `compact_objects()` request with an empty `compaction_map_t`. Though, I think we'll have to change `replace_objects_update`:
https://github.com/redpanda-data/redpanda/blob/79e74eb1898e9434fb43cb0ab8bec07c7b5f7950/src/v/cloud_topics/level_one/metastore/state_update.h#L162-L165

to contain an `std::optional<...> compaction_updates` in order to disambiguate a compaction request from a replace objects request here: https://github.com/redpanda-data/redpanda/blob/79e74eb1898e9434fb43cb0ab8bec07c7b5f7950/src/v/cloud_topics/level_one/metastore/state_update.cc#L408-L410

andrwng commented on src/v/cloud_topics/level_one/compaction/committer.h:
```diff
@@ -30,188 +30,203 @@ namespace cloud_topics::l1 {
+        // before returning the result.
+        ss::future<expected_t> upload_file(file_and_md_info);
+
+        // Attempts to upload & commit all updates currently in
```
nit: it doesn't seem like this loop actually commits to the metastore, if that's what this is saying. If that's not what commit means, maybe can we clarify the verbiage so it's clearer what commit means?

andrwng commented on src/v/cloud_topics/level_one/compaction/committer.h:
```diff
@@ -30,188 +30,203 @@ namespace cloud_topics::l1 {
+        // Starts the backgrounded committing loop.
+        void start_upload_loop();
+
+        // Attempts to create a metastore object using the `metadata_builder`,
+        // put the object into cloud storage, and finalize the object. Returns
+        // an error if any step of that process should fail.
+        ss::future<expected_t> do_upload(
+          staging_file*,
+          object_builder::object_info,
+          metastore::object_metadata::ntp_metadata);
```
nit: what's the point of separating out do_upload() and do_finalize() from upload_file() and finalize()?

For the most part it seems like they only differ by a couple important lines, but each is called just once. So if there isn't an important distinction, could we inline the do_* calls so it's easier to reason about where the uploading and finalizing is happening (less code to search)?

WillemKauf replied:
Both `do_upload()` and `do_finalize()` have a lot of branching/unhappy paths, and we need to unconditionally call `file->remove()` & `_committer->finalize_job(_id);` (respectively) after. The logic is easier to parse, IMO, when they are split off into separate functions, and the unconditional function calls are present within  `upload_file()`/`finalize()`.

andrwng commented on src/v/cloud_topics/level_one/compaction/committer.cc:
```diff
@@ -10,101 +10,471 @@
+        return;
+    }
+
+    job_it->second.get()->cancel_job();
```
Doesn't seem like this is used, and I think that means we don't need _job_id_by_tp?

WillemKauf replied:
Yeah, not at present. I'll remove it for now.

andrwng commented on src/v/cloud_topics/level_one/compaction/committer.cc:
```diff
@@ -10,101 +10,471 @@
+      "Awaiting completion of uploads for job {}, tidp {}",
+      _id,
+      _tp);
+
```
Is there a guarantee that _staging_file_and_md_infos is non-empty here? Do we need to upload_some() before this?

andrwng commented on src/v/cloud_topics/level_one/compaction/committer.h:
```diff
@@ -25,63 +30,220 @@ namespace cloud_topics::l1 {
+        void upload_some();
+
+        // Awaits all inflight uploads. It is expected that
+        // `all_uploads_inflight()` has been checked before calling this
```
Where does this happen? Maybe something we can dassert on?

WillemKauf replied:
Ah oops. There was a missing `ss::condition_variable` that got removed in the rework.

Nothing remarkable to `dassert()` on here, since we have two potential conditions when entering this function:

1. `all_uploads_inflight()` is `true` and we now have to wait on `_upload_cv`.
2. We are in the middle of a shutdown

WillemKauf replied:
Force pushed with this commit diff: https://github.com/redpanda-data/redpanda/compare/206b268e5332986221ab8a56729fafab13549cce..a322806add3d0da9acbf56ec819620fc2c02862f

WillemKauf committed:
`ct/l1`: add `object_metadata_builder::is_empty()`

A new function to help determine if the built object metadata update
is empty. We can use this to avoid pushing a no-op/bad update to the
`metastore`.

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/77744-->
<!--10d68860-ac02-4384-939c-f49e62928ff3-->
<!--e8bb9870-edd6-4553-baf0-c4413e659802-->
<!--643a8383-a33e-4112-aedb-0e5bf99da192-->
<!--c7c1daa8-40f1-4a24-8afc-a20ef4b8d7a3-->
<!--a762559e-d502-45e6-acea-743628e4e33d-->
<!--title-->
## Retry command for [Build#77744](https://buildkite.com/redpanda/redpanda/builds/77744)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
skip-redpanda-build
skip-units
skip-rebase
tests/rptest/tests/license_enforcement_test.py::LicenseEnforcementTest.test_license_enforcement@{"clean_node_after_recovery":true,"clean_node_before_recovery":true}
tests/rptest/tests/cluster_features_test.py::FeaturesNodeJoinTest.test_old_node_join
tests/rptest/tests/cluster_features_test.py::FeaturesSingleNodeUpgradeTest.test_upgrade
tests/rptest/tests/license_enforcement_test.py::LicenseEnforcementTest.test_license_enforcement@{"clean_node_after_recovery":true,"clean_node_before_recovery":false}
tests/rptest/tests/cluster_features_test.py::FeaturesMultiNodeUpgradeTest.test_rollback
tests/rptest/tests/license_enforcement_test.py::LicenseEnforcementTest.test_license_enforcement@{"clean_node_after_recovery":false,"clean_node_before_recovery":false}
tests/rptest/tests/cluster_features_test.py::FeaturesMultiNodeUpgradeTest.test_upgrade
tests/rptest/tests/license_enforcement_test.py::LicenseEnforcementTest.test_license_enforcement@{"clean_node_after_recovery":false,"clean_node_before_recovery":true}
```
<!--/body-->

WillemKauf commented:
/ci-repeat 1

andrwng commented on src/v/cloud_topics/level_one/compaction/committer.cc:
```diff
@@ -10,101 +10,479 @@
         }
+    }
+
+    _upload_cv.signal();
```
nit: maybe call this something like _last_upload_scheduled? It's otherwise a bit confusing the difference between this and _upload_sem
